Renaming and general clean-up for new init

ChrisMzz opened the pull request:
Just created a new branch to kickstart the changes. The very first change I made was removing every instance of the word "cell" that isn't "cell cycle". We discussed with Léo that "cell cycle" is still ambiguous, but I feel like every other instance of the word cell could be used interchangeably with "node".

For example, this `cell` was changed to `node` : 
```py
roots = [cell for cell in roots if self.image_label[cell] != 1]
```

Let's discuss additional changes to be made and commit them to this branch if that's ok with you both.

ChrisMzz committed:
find-replace cell to node when relevant

ChrisMzz commented:
About `all_tracks`, `get_tracks`, and `get_all_tracks` : 

There is a redundancy between `get_tracks` and `get_all_tracks`, since the two lines below produce the exact same code (the code is identical in both cases) : 
```py
lT.get_tracks(lT.roots)
lT.get_all_tracks()
```
I know that `get_tracks` supports `roots=None` (in which case we currently return `lT.get_all_tracks()`) but we could just set `roots` to `lT.roots` in that case and we'd achieve the exact same result.

From this change, in the `all_tracks` property :
```py
    @property
    def all_tracks(self) -> list[list[int]]:
        if not hasattr(self, "_all_tracks"):
            self._all_tracks = self.get_all_tracks()
        return self._all_tracks
```
we would just replace `get_all_tracks` by `get_tracks`.


What do you two think @leoguignard @BadPrograms ? I know we also discussed `all_tracks` should become a `dynamic_property` but it made `_all_tracks` set itself to `None` on initialisation, so I might need help making that change.

ChrisMzz committed:
find-replace 'track' to 'chain'

ChrisMzz committed:
find-replace (cell) cycle to chain

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -440,20 +440,20 @@ def __setstate__(self, state):
 
     def _get_height(self, c: int, done: dict) -> float:
-        """Recursively computes the height of a cell within a tree * a space factor.
+        """Recursively computes the height of a node within a tree * a space factor.
```
Actually here can we change `*` by `times`

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -646,20 +646,20 @@ def stroke_color(x):
-                x2, y2 = positions[c_cycle[-1]]
+                c_chain = self.get_chain(curr)
+                x1, y1 = positions[c_chain[0]]
+                x2, y2 = positions[c_chain[-1]]
```
Don't change it but here the `c` in `c_cycle` is for cell. Therefore, changing to `chain` instead of `c_chain` would have been slightly better. Again, do not change that :)

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -1158,32 +1158,32 @@ def get_predecessors(
         while (
             acc != depth
-            and start_time < self._time[unconstrained_cycle[0]]
+            and start_time < self._time[unconstrained_chain[0]]
             and (
-                self._predecessor[unconstrained_cycle[0]] != ()
+                self._predecessor[unconstrained_chain[0]] != ()
                 and (  # Please dont change very important even if it looks weird.
                     len(
                         self._successor[
-                            self._predecessor[unconstrained_cycle[0]][0]
+                            self._predecessor[unconstrained_chain[0]][0]
                         ]
                     )
                     == 1
                 )
             )
         ):
```
@BadPrograms
Wouldn't that be more readeable?
```python
while (
    acc != depth
    and start_time < self._time[first_element := unconstrained_chain[0]]
    and (pred := self._predecessor[first_element]) != ()
    and len(self._successor[pred]) == 1
):
```
(real question)

BadPrograms replied:
Honestly, the Walrus operator seems strange to me. I don't know if it will be more readable this way.

leoguignard replied:
Fair enough for the walrus operator, that being said, that might still be better:
```python
while (
    acc != depth
    and start_time < self._time[unconstrained_chain[0]]
    and (self._predecessor[unconstrained_chain[0]]) != ()
    and len(self._successor[self._predecessor[unconstrained_chain[0]]]) == 1
):
```

leoguignard replied:
basically removing the unnecessary parenthesis

BadPrograms replied:
Practically it's the same, right? I mean runtime. However, this looks better I agree.

leoguignard replied:
Yes, it's just for readability

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -1260,10 +1260,10 @@ def get_cell_cycle(
-        if not hasattr(self, "_all_tracks"):
-            self._all_tracks = self.get_all_tracks()
-        return self._all_tracks
+    def all_chains(self) -> list[list[int]]:
+        if not hasattr(self, "_all_chains"):
+            self._all_chains = self.get_all_chains()
+        return self._all_chains
```
let's add a docstring here maybe

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -1300,8 +1300,8 @@ def parenting(self):
     def get_all_branches_of_node(
         self, node: int, end_time: int = None
     ) -> list[list[int]]:
-        """Computes all the tracks of the subtree spawn by a given node.
-        Similar to get_all_tracks().
+        """Computes all the chains of the subtree spawn by a given node.
+        Similar to get_all_chains().
```
why don't we rename that function `get_all_chains_from_node`?

ChrisMzz replied:
I have a commit ready on my end that changes all instances of "branch" to "chain" when relevant, is now a good time to push that?

leoguignard replied:
yes please do so

leoguignard replied:
Name propositions:
`get_all_subtree_chains`
`get_all_chains_from_subtree`


Behaviour:
```text
1
|
2
|
3
|------|
4      7
|      |
5      8
|      |
6      9
```

```python
c = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
```

```python
get_all_subtree_chains(2) # -> [[2, 3], [4, 5, 6], [7, 8, 9]]
```

leoguignard replied:
Let's have a function `get_node_chain`:
```python
def get_node_chain(self, node: int):
     # Former get_cycle function
```

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -1313,42 +1313,42 @@ def get_all_branches_of_node(
         -------
         list of list of int
-            list of lists containing track cell ids
+            list of lists containing chain node ids
```
maybe better:
list of chains, chains are list of node ids.

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -1313,42 +1313,42 @@ def get_all_branches_of_node(
-                branches += [track]
-                to_do += self._successor[track[-1]]
+            chain = self.get_successors(current, end_time=end_time)
+            # if len(chain) != 1 or self._time[current] <= end_time:
```
let's remove that comment

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -1260,10 +1260,10 @@ def get_cell_cycle(
         ] + self.get_successors(x, depth_succ, end_time=end_time)
 
     @property
```
Is the change to `@dynamic_property` for another PR?

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -1313,42 +1313,42 @@ def get_all_branches_of_node(
-    def get_all_tracks(self, force_recompute: bool = False) -> list[list[int]]:
-        """Computes all the tracks of a given lineage tree,
-        stores it in `self.all_tracks` and returns it.
+    def get_all_chains(self, force_recompute: bool = False) -> list[list[int]]:
```
I think that since this method litteraly returns `self.all_chains` it should be made a private method.
```python
def _compute_all_chains(...):
```

leoguignard commented on src/LineageTree/lineageTree.py:
```diff
@@ -1313,42 +1313,42 @@ def get_all_branches_of_node(
 
-    def get_tracks(self, roots: list = None) -> list[list[int]]:
-        """Computes the tracks given by the list of nodes `roots` and returns it.
+    def get_chains(self, roots: list = None) -> list[list[int]]:
```
how is that different from `self.get_all_branches_of_node`? I mean I see that one has `end_time` and the other has not but couldn't these two merge into one method?

leoguignard replied:
Also, if `all_chains` is computed correctly, we should be able to compute `get_chains` in $O(n)$, $n$ being the number of chains

leoguignard replied:
Let's discuss these functions together

BadPrograms replied:
> Also, if `all_chains` is computed correctly, we should be able to compute `get_chains` in O ( n ) , n being the number of chains

Can you explain this?

leoguignard replied:
If the list of chains is written DFS style then you just have to go through the list of chains, as soon as you find the cell you are looking for you start appending the chains until you find a chain that starts at an earlier time than your initial chain

leoguignard replied:
Something like that should work I think
```python
all_chains = self.all_chains
for n in roots:
    starting_node = self.get_predecessors(n)[0]
    found = False
    done = False
    starting_time = self.time[n]
    i = 0
    output_chains = []
    while not done and i < len(all_chains):
        found = chain[i][0] == starting_node
        if found:
            done = self.time[chain[i]] < starting_time
            if not done:
                output_chains.append(chain[i])    
        i += 1
```

leoguignard replied:
I guess someone could do it in one go actually, by keeping a list of `done`, `found`, `starting_time` and `output_chains` and having
```python
while not all(done) and ...:
    ...
```

BadPrograms commented:
Also, the get_successors and get_predecessors may be changed to get_successor_chain and get_predecessor_chain to make more sense.

BadPrograms commented on src/LineageTree/lineageTree.py:
```diff
@@ -1158,32 +1158,32 @@ def get_predecessors(
-        return cycle
+        return chain
 
     def get_successors(
```
The get_successors and get_predecessors funcs may be renamed to get_successor_chain and get_predecessor_chain to make more sense.

leoguignard replied:
I don't think so, a chain is defined as starting either after a division or at a root and ending either at a division or at a leaf.

`get_successors` gives the list of successors and `get_predecessors` gives a list of predecessors

BadPrograms replied:
So if a chain is [1,2,3,4,5], how should the nodes [2,3,4] be called?

leoguignard replied:
a subchain?

leoguignard replied:
or a path, or something else. Do we need to name that?

leoguignard replied:
I mean, if you can't easily define it then don't name it?

ChrisMzz replied:
I'm pretty sure what `get_successors` and `get_predecessors` return is not actually the successors/predecessors of the given node. I agree that the names should be changed at the bare minimum, because one would expect :
```py
lT = lineageTree(successor{0:(1,), 1:(2,), 2:(3,4)})
lT.get_successors(2)
```
to give `[3,4]`, and yet it outputs `[2]`.

BadPrograms replied:
> ```python
> lT = lineageTree(successor{0:(1,), 1:(2,), 2:(3,4)})
> lT.get_successors(2)
> ```

This is intended.
Also I understand the chain as a any path in a graph that does not contain a critical point (critical points are roots, divisions and leaves). So the subchain of a chain should be a chain.

ChrisMzz committed:
find-replace branch to chain

ChrisMzz commented:
Details of latest commit ([b9cc022](https://github.com/GuignardLab/LineageTree/pull/68/commits/b9cc02200df5294fed193d4fb94c5719eb3318bb)) :
- made `all_chains` a `@dynamic_property`
- fixed docstrings
- changed `get_sub_tree` to `get_sub_tree_nodes`
- changed `get_all_chains` to `_compute_all_chains` (and made it only be called in the property)
- changed `get_chains_from_node` to `get_all_subtree_chains`
- changed `get_chain` to `get_node_chain`
- changed `plot_node` to `plot_sub_tree`

ChrisMzz committed:
see [PR#68 comment](https://github.com/GuignardLab/LineageTree/pull/68#issuecomment-2847356131)

leoguignard commented on src/LineageTree/tree_styles.py:
```diff
@@ -231,7 +231,7 @@ def delta(self, x, y, corres1, corres2, times1, times2):
     def get_norm(self) -> int:
         return (
-            len(self.lT.get_sub_tree(self.root, end_time=self.end_time))
+            len(self.lT.get_sub_tree_nodes(self.root, end_time=self.end_time))
```
We need to agree on `sub_tree` or `subtree`

BadPrograms replied:
subtree, to have less words

ChrisMzz replied:
I agree with `subtree`. I'll make the change everywhere in a second.

leoguignard reviewed (changes requested):
Still some changes. The main one being the reimplementation of `get_chains`
The implementation I suggested might work but I did it from scratch so might have errors.
It is crucial to double check that it works as intended!!

ChrisMzz commented:
> Still some changes. The main one being the reimplementation of `get_chains` The implementation I suggested might work but I did it from scratch so might have errors. It is crucial to double check that it works as intended!!

Indeed, as it was the only change that re-implements a method I preferred leaving it for a separate commit. I'd already tested some things out from the implementation you suggested on Friday, so I'll continue this now.

ChrisMzz committed:
find-replaced 'sub_tree' to 'subtree'

ChrisMzz committed:
reimplemented `get_chains` and ordered `all_chains`

ChrisMzz commented:
With [2cf47db](https://github.com/GuignardLab/LineageTree/pull/68/commits/2cf47dbeb8e0117236baa37db1b488bd6c2d8745) I reimplemented `get_chains` and tested it on the following examples : 
```py
from LineageTree import lineageTree
test_lT = lineageTree(successor={1: (2,), 2: (3, 100), 100: (101,), 0: (1,), 10: (0,), 5: (), 3: (), 4: (88,89), 88:(90,), 101:(102,103), 103:(104,)})
test_lT2 = lineageTree(
    successor={1: (2,), 2: (3, 100), 100: (101,), 0: (1,), 10: (0,), 5: (), 3: (), 4: (), 101: ()}, 
    time={10:0, 0:1, 1:2, 2:3, 3:4, 100:4, 101:5, 5:2, 4:20}
)
```
I noticed an inconsistency in the chains ordering which led to `test_lT.get_chains()` not giving the chains in the same order as `test_lT.get_chains(test_lT.roots)`. Fixing this would require [L.1364](https://github.com/GuignardLab/LineageTree/blob/2cf47dbeb8e0117236baa37db1b488bd6c2d8745/src/LineageTree/lineageTree.py#L1364) in `get_chains` to become : 
```py
roots = sorted(list(roots), key=self.time.get)[::-1]
```
so that roots are treated in the same order as they are in `_compute_all_chains`. The fix is a bit ugly but we can't change the ordering in `all_chains` because of the DFS ordering requirement.

If this is fine with you @leoguignard I'll commit this ASAP and if I'm not mistaken that only leaves us with the "ambiguous names" part of the list of changes.

ChrisMzz committed:
fix new `get_all_chains` implementation + docstrings

ChrisMzz committed:
remove `scale_embryo`

ChrisMzz committed:
renamed `to_simple_graph` to `_create_dict_of_plots`

ChrisMzz commented:
## Changelog and comments

### Chain
The terms "cell cycle", "branch" and "track" have been removed as to introduce the term "chain".
A chain is a path that begins either at a root or at the successor of a node which has more than one successor, and ends in a leaf or a node with two successors.
This change in the framework led to the following changes : 
- `get_tracks` was changed to `get_chains`
- `get_all_tracks` was changed to `_compute_all_chains`
- `all_tracks` was changed to `all_chains` (and made into a `dynamic_property`)
- `get_all_branches_of_node` was changed to `get_all_chains_of_subtree` (not `get_all_subtree_chains`, changed for clarity)
- `get_cell_cycle` was changed to `get_node_chain`
- `add_branch` was changed to `add_chain`

On a side note, the `get_chains` method was re-implemented to make proper use of the `all_chains` property, which is now the only instance of `_compute_all_chains` (which is why it was made internal)

### Minor edits

I did my best to group all of these together but I may have forgotten some, let me know if you think of any others : 
- all instances of `sub_tree` were changed to `subtree`.
- `plot_node` was changed to `plot_subtree`
- `scale_embryo` was removed
- `to_simple_graph` was changed to `_create_dict_of_plots`

BadPrograms commented:
Perfect! I think we are ready to merge!

ChrisMzz committed:
change `get_all_subtree_chains` to `get_all_chains_of_subtree`
